### ISSUE TITLE
Possible ModalBox Accessibility Improvements

### ISSUE THREAD

**Issue by markcaron** [patternfly/patternfly-next#2047]
I'm curious to the DOM order of the close button in the ModalBox, and I was hoping this could be a discussion so we could arrive at the best possible modal experience.

Currently, the close button is the very first element in the Dialog Box DOM which makes for a slightly weird experience in VoiceOver—after the modal opens and reads the aria-label on the `dialog` it focuses immediately on the close button and announces it next before telling the user how many items are in the dialog and then moving on to read the rest of the content in modal (if the user allows it to continue). This isn't a huge deal, it's just odd and unexpected, IMO. I think it'd be better if the close button was added just after the body content (`pf-c-modal-box__body`) and before the footer (`pf-c-modal-box__footer`, with the other action buttons), and I believe it'd make it more consistent with the NVDA and JAWS experiences too. And, nothing would change visually either—which is awesome!

It's also worth noting, too, that if the focus is placed on the first item (the close button), then _if_ the modal contained a form, it wouldn't place focus on the first element that the user would probably be the most interested in (the first input perhaps) like the [WAI example](https://www.w3.org/TR/wai-aria-practices/examples/dialog-modal/dialog.html). This seems like a bit of an issue to me. Does the close button have priority over the actionable items within the modal (like forms)?

Looking at other modal patterns, like [The A11Y Project's](https://a11yproject.com/patterns/#modal-windows), the close button either comes toward the end of the modal ([example one](http://scottaohara.github.io/accessible_modal_window/), [example two](https://codepen.io/BeyondHyper/pen/NREjpP) and the [WAI example](https://www.w3.org/TR/wai-aria-practices/examples/dialog-modal/dialog.html)) _**or**_ the focus is placed on the modal dialog itself (in the ["Launchy" example](https://codepen.io/svinkle/pen/pjVepz) which I'm not a total fan of, but is better than immediately focusing on the close button).

I think moving the DOM order of the close button could be a super simple change with zero visual impact and added screenreader consistency. A win:win!

Another thing WAI and all the other examples do is to use the `aria-labelledby` attribute to "name" the dialog via the first heading. It may be splitting hairs, but using `aria-label` is slightly different (implicit vs explicit). I'm not sure how it works in the code in the backend, but if there is a chance that the aria-label and the modal header are different, then that could be an issue. Using the `aria-labelledby` will never be different, and it does take precedence over `aria-label`, but it does require generating an ID (which may be your reason for going with label). Not something that's really necessary to change, just food for thought.

Lastly, the ["Incredible Accessible Modal Window"](https://codepen.io/BeyondHyper/pen/NREjpP) uses `aria-describedby` to further explain to the user that the modal can be closed and cancelled via "Esc". This may be an issue with translations, but I just wanted to put this out there as a good/helpful feature.

@jgiardino, thoughts? Thanks for reading my crazy ruminations.

**Comment by mcoker** [comment]
cc @jessiehuff

**Comment by jessiehuff** [comment]
I think this could be a valid discussion, but it's a tough one for me to have a very strong opinion on. I believe that we have initial focus on the close button since it's the "least destructive" action for a user. When talking to Jenn about it, she thought it'd be necessary to test in all screen readers to confirm what the experience is when the dialog is announced and the differences between approaches to really have an opinion on this. This is one of those issues that would require some effort to test and capture findings to be able to have a conclusion on the best solution that's possible for PF. However, we're not sure these changes would provide a significant benefit to the user and/or benefits that might be subjective. So I'm ok keeping this open, but I think it's low priority compared to some of the other accessibility issues currently open.

**Comment by thatblindgeye** [comment]
@markcaron it seems like this issue can be closed out.

- When opening a Modal with VoiceOver, the announcement I get is "Dialog with 4 items, close button". This is in Chrome, Firefox, and Safari (albeit with some slight differences in exact announcement -- Chrome says the aforementioned quote, Firefox goes on to announce the rest of the content in the modal, and Safari announces the modal title first).
- I feel like moving the close button in the DOM (but keeping it visually in the same spot) may introduce a different sort of confusion. Visually the expectation might be that the close button receives focus first, but it doesn't receive focus until after most everything else in the Modal, and possibly moves focus in an odd path (header > body > close button then other footer content or footer content then back to the top of the modal to the close button). This is a bit subjective, though, and there could be an argument that even our current DOM order could still be improved since the close button (visually after the heading horizontally) is rendered before the heading in the DOM, so we're going from end to start.
- React now has an `elementToFocus` prop that allows customizing what element receives focus upon the modal opening, which would address scenarios as you noted where a form input might need to be the initially focused item in a modal. Note that I opened [an issue to add content to our a11y docs](https://github.com/patternfly/patternfly-org/issues/4416) to cover this prop as well.
- Modal currently allows all 3 aria labeling attributes - labelledby, label, and describedby. Our [basic modal example](https://www.patternfly.org/components/modal#basic-modals) shows labelledby and describedby being used together, while the [no header or footer modal example](https://www.patternfly.org/components/modal#no-header-or-footer) shows a use case for using aria-label.
- If wanted a consumer could hook up instructions related to Escape closing the modal via aria-describedby, though I'd argue it either a) may not be necessary as users may be used to the typical interactions of a modal, or b) that hidden text in the example shouldn't really be hidden at all as it could benefit users without low enough vision as well.

With all that said, I will be closing this out for now, but feel free to comment or open a followup issue if there is still a strong desire for updates to be made to the Modal component.
